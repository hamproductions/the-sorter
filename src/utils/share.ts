import type { FilterType } from '~/components/sorter/CharacterFilters';

export const serializeData = async (data: Record<string, unknown>) => {
  const compress = (await import('lz-string')).compressToEncodedURIComponent;
  return compress(JSON.stringify(data));
};

export const addPresetParams = (
  params: URLSearchParams,
  filters: FilterType,
  isSeiyuu: boolean
) => {
  for (const key of ['series', 'school', 'units'] as const) {
    const list = filters?.[key];
    if (list && list?.length > 0) {
      list.forEach((item) => params.append(key, item));
    }
  }
  if (isSeiyuu) {
    params.append('seiyuu', isSeiyuu.toString());
  }
  return params;
};
