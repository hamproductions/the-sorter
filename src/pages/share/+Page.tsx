import { Suspense, lazy } from 'react';
import { useTranslation } from 'react-i18next';
import { decompressFromEncodedURIComponent } from 'lz-string';
import { Link } from '../../components/ui/styled/link';
import { Text } from '../../components/ui/styled/text';
import { useData } from '../../hooks/useData';
import type { Character } from '../../types';
import school from '../../../data/school.json';
import series from '../../../data/series.json';
import units from '../../../data/units.json';
import { Container, Stack } from 'styled-system/jsx';
import { ResultsView } from '~/components/results/ResultsView';
import { getFilterTitle } from '~/utils/filter';

import { Button } from '~/components/ui/styled/button';
import { Metadata } from '~/components/layout/Metadata';
import type { TierListSettings } from '~/components/results/TierList';
import { useDialogData } from '~/hooks/useDialogData';

const CharacterInfoDialog = lazy(() =>
  import('../../components/dialog/CharacterInfoDialog').then((m) => ({
    default: m.CharacterInfoDialog
  }))
);

export function Page() {
  const data = useData();
  const { t, i18n } = useTranslation();

  const {
    data: showCharacterInfo,
    isOpen: isShowCharacterInfo,
    setData: setShowCharacterInfo
  } = useDialogData<Character>();

  const params = new URLSearchParams(import.meta.env.SSR ? '' : location.search);
  const urlSeiyuu = params.get('seiyuu');

  const urlSeries = params.getAll('series');
  const urlSchool = params.getAll('school');
  const urlUnits = params.getAll('units');
  const urlData = params.get('data');

  const {
    results,
    ...shareDisplayData
  }: {
    title: string;
    description?: string;
    tierlist?: TierListSettings;
    results: string[][];
    tab?: 'default' | 'table' | 'grid' | 'tier';
  } = JSON.parse(urlData !== null ? decompressFromEncodedURIComponent(urlData) : '{}') ?? {};

  const seiyuu = urlSeiyuu === 'true';

  const filters = {
    series: urlSeries.filter((s) => Object.keys(series).includes(s)),
    school: urlSchool.filter((s) => Object.keys(school).includes(s)),
    units: urlUnits.filter((unitId) => units?.some((s) => s.id === unitId))
  };

  const titlePrefix = getFilterTitle(filters, data, i18n.language) ?? t('defaultTitlePrefix');
  const title = t('title', {
    titlePrefix
  });

  const getShareUrl = () => {
    if (import.meta.env.SSR) return '/';
    const params = new URLSearchParams(location.search);
    params.delete('data');
    return `${location.origin}${import.meta.env.PUBLIC_ENV__BASE_URL ?? '/'}?${params.toString()}`;
  };

  return (
    <>
      <Metadata title={title} helmet />
      <Container zIndex="1" flex={1} w="full" py={4} px={4}>
        <Stack alignItems="center" w="full">
          <Text textAlign="center" fontSize="3xl" fontWeight="bold">
            {title}
          </Text>
          <Text textAlign="center">{t('description')}</Text>
          <Link href={getShareUrl()}>
            <Button>{t('share.create_your_own')}</Button>
          </Link>
          {data?.length && results?.length > 0 && (
            <>
              <ResultsView
                titlePrefix={titlePrefix}
                charactersData={data}
                isSeiyuu={seiyuu}
                shareDisplayData={shareDisplayData}
                onSelectCharacter={(c) => setShowCharacterInfo(c)}
                readOnly
                w="full"
                order={results}
              />
              <Link href={getShareUrl()}>
                <Button>{t('share.create_your_own')}</Button>
              </Link>
            </>
          )}
        </Stack>
      </Container>
      <Suspense>
        <CharacterInfoDialog
          character={showCharacterInfo}
          isSeiyuu={seiyuu}
          open={isShowCharacterInfo}
          onOpenChange={(e) => {
            if (!e.open) {
              return setShowCharacterInfo(undefined);
            }
          }}
        />
      </Suspense>
    </>
  );
}
