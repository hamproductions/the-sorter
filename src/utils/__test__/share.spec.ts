import { describe, expect, it } from 'vitest';
import { addPresetParams, serializeData } from '../share';

describe('Share Utils', () => {
  describe('addPresetParams function', () => {
    it('Create create parameters for preset', () => {
      expect(
        addPresetParams(
          new URLSearchParams(),
          { series: ['hoge'], school: ['fuga'], units: ['1', '2', '3', '4'] },
          false
        ).toString()
      ).toEqual('series=hoge&school=fuga&units=1&units=2&units=3&units=4');
      expect(
        addPresetParams(
          new URLSearchParams(),
          { series: ['hoge'], school: ['fuga'], units: ['1', '2', '3', '4'] },
          true
        ).toString()
      ).toEqual('series=hoge&school=fuga&units=1&units=2&units=3&units=4&seiyuu=true');
    });
  });

  describe('serializeData function', () => {
    it('Serialize Data with lz compression algorithm', async () => {
      expect(await serializeData({ data: 'hoge' })).toMatchSnapshot();
    });
  });
});
