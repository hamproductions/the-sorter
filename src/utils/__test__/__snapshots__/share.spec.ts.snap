// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Share Utils > serializeData function > Serialize Data with lz compression algorithm 1`] = `"N4IgJghgLhIFwgBYHsDmBTEBfIA"`;
